python/ray/tune/examples/xgboost_dynamic_resources_example.py

```diff
@@ -13,13 +13,73 @@
 from ray.tune import Trainable
 from ray.tune.resources import Resources
 from ray.tune.utils.placement_groups import PlacementGroupFactory
-from ray.tune.suggest.basic_variant import BasicVariantGenerator
 from ray.tune.trial import Trial
 from ray.tune import trial_runner
+from ray.tune.integration.xgboost import TuneReportCheckpointCallback
 
+CHECKPOINT_FILENAME = "model.xgb"
 
-# Dynamic resource allocation is currently only possible
-# with Trainable (class) API
+
+def get_best_model_checkpoint(analysis):
+    best_bst = xgb.Booster()
+    try:
+        with open(analysis.best_checkpoint, "rb") as inputFile:
+            _, _, raw_model = pickle.load(inputFile)
+        best_bst.load_model(bytearray(raw_model))
+    except IsADirectoryError:
+        best_bst.load_model(
+            os.path.join(analysis.best_checkpoint, CHECKPOINT_FILENAME))
+    accuracy = 1. - analysis.best_result["eval-logloss"]
+    print(f"Best model parameters: {analysis.best_config}")
+    print(f"Best model total accuracy: {accuracy:.4f}")
+    return best_bst
+
+
+# FUNCTION API EXAMPLE
+
+
+# our train function needs both checkpoint_dir and new_resources
+# to work with ResourceChangingScheduler
+def train_breast_cancer(config: dict, checkpoint_dir=None, new_resources=None):
+    # This is a simple training function to be passed into Tune
+    # Load dataset
+    data, labels = sklearn.datasets.load_breast_cancer(return_X_y=True)
+    # Split into train and test set
+    train_x, test_x, train_y, test_y = train_test_split(
+        data, labels, test_size=0.25)
+    # Build input matrices for XGBoost
+    train_set = xgb.DMatrix(train_x, label=train_y)
+    test_set = xgb.DMatrix(test_x, label=test_y)
+
+    # Checkpointing needs to be set up in order for dynamic
+    # resource allocation to work as intended
+    xgb_model = None
+    if checkpoint_dir:
+        xgb_model = xgb.Booster()
+        xgb_model.load_model(os.path.join(checkpoint_dir, CHECKPOINT_FILENAME))
+
+    if new_resources is None:
+        config["nthread"] = 1
+    else:
+        config["nthread"] = int(new_resources.head_cpus)
```

Comment:
What I dislike here is that we now have two separate locations where we deal with resource configurations. Previously we would set resources only in `resources_per_trial` within `tune.run()` and implicitly assumed that the function trainable would be aware of that or self-detect assigned resources. Now we explicitly fetch resource information from an argument.

I was wondering if we should rather have global lookup functions like `tune.get_trial_resources()` (similar to the existing `tune.get_trial_id()`. That way each trial could fetch runtime information as needed and we wouldn't need to clutter the function API interface

Reply:
That's a very good point. Let me look into doing that instead.